File: src/main/java/validparenthesis/ValidParenthesis.java
package validparenthesis;

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Map;

public class ValidParenthesis {
    boolean isValid(String s) {
        if (s == null) {
            return false;
        }
        var stack = new ArrayDeque<Character>();
        var matching = Map.of(
                ')', '(',
                '}', '{',
                ']', '['
        );
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            if (c == '(' || c == '{' || c == '[') {
                stack.push(c);
            } else {
                if (stack.isEmpty() || stack.pop() != matching.get(c)) {
                    return false;
                }
            }
        }
        return stack.isEmpty();
    }

    void main(String[] args) {
        System.out.println(isValid("([]{})"));
    }
}

File: src/main/kotlin/validParenthesis/ValidParenthesis.kt
package validparenthesis

// #### Valid Parentheses:
// Given a string containing just the characters '(', ')', '{', '}', '[' and ']',
// determine if the input string is valid.

private val CLOSING_TO_OPENING = mapOf(')' to '(', '}' to '{', ']' to '[')

fun isValidString(s: String): Boolean {
    val stack = ArrayDeque<Char>()

    for (char in s) {
        when (char) {
            in CLOSING_TO_OPENING.values -> stack.addLast(char)
            in CLOSING_TO_OPENING.keys ->
                if (stack.removeLastOrNull() != CLOSING_TO_OPENING[char]) return false
        }
    }

    return stack.isEmpty()
}

fun main() {
    println(isValidString("[]()"))   // true
    println(isValidString("{[()]}")) // true
    println(isValidString("([)]"))   // false
}
